### PR TITLE
add ExtendedNodeStats interface incl. default implementation

### PR DESCRIPTION
**Pull request by seut**

### CODE REVIEW

File: sql/src/main/java/io/crate/stats/ExtendedFsStats.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            this.diskWrites = diskWrites;
+            this.diskReadBytes = diskReadBytes;
+            this.diskWriteBytes = diskWriteBytes;
+        }
+
+        public BytesRef path() {
+            return path;
+        }
+
+        public BytesRef dev() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/stats/ExtendedFsStats.java]: annotate as Nullable please

File: sql/src/main/java/io/crate/stats/ExtendedFsStats.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            if (existing < 0) {
+                return added;
+            }
+            return existing + added;
+        }
+
+    }
+
+    Info total;
+    Info[] infos;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/stats/ExtendedFsStats.java]: can infos be marked as final?

File: sql/src/main/java/io/crate/stats/ExtendedNetworkStats.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ */
+
+package io.crate.stats;
+
+public class ExtendedNetworkStats {
+
+    long timestamp;
+    Tcp tcp = null;
+
+    public ExtendedNetworkStats() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/stats/ExtendedNetworkStats.java]: is this ctor in use?

File: sql/src/main/java/io/crate/stats/ExtendedOsStats.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.util.concurrent.TimeUnit;
+
+public class ExtendedOsStats {
+
+    long timestamp;
+    long uptime = -1;
+    double[] loadAverage = new double[0];
+    Cpu cpu;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/stats/ExtendedOsStats.java]: can cpu be marked as final?

File: sql/src/main/java/io/crate/stats/ExtendedOsStats.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    public Cpu cpu() {
+        return cpu;
+    }
+
+    public static class Cpu {
+
+        short sys = -1;
+        short user = -1;
+        short idle = -1;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/stats/ExtendedOsStats.java]: those attributes can also be final right?

File: sql/src/main/java/io/crate/stats/ExtendedProcessCpuStats.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+package io.crate.stats;
+
+import org.elasticsearch.common.unit.TimeValue;
+
+public class ExtendedProcessCpuStats {
+
+    short percent = -1;
+    long sys = -1;
+    long user = -1;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/stats/ExtendedProcessCpuStats.java]: same here

File: sql/src/main/java/io/crate/stats/ExtendedProcessCpuStats.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        return percent;
+    }
+
+    /**
+     * Get the Process cpu kernel time.
+     * <p/>
+     * <p>Supported Platforms: All.
+     */
+    public TimeValue sys() {
+        return new TimeValue(sys);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/stats/ExtendedProcessCpuStats.java]: would it make sense to create the timeValue instance in the ctor in order to do it only once?

File: sql/src/main/java/io/crate/stats/StatsModule.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    private final Settings settings;
+    private final Map<String, Class<? extends ExtendedNodeStats>> extendedStatsTypes = new HashMap<>();
+
+    public StatsModule(Settings settings) {
+        this.settings = settings;
+        addExtendedStatsType("none", ZeroExtendedNodeStats.class);
+    }
+
+    public void addExtendedStatsType(String type, Class<? extends ExtendedNodeStats> clazz) {
+        if (extendedStatsTypes.containsKey(type)) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/stats/StatsModule.java]: you could also check the return value of put which saves an extra containsKey call =)

File: sql/src/main/java/io/crate/stats/ExtendedOsStats.java (hunk 5 above)
Comment (seut) [inline sql/src/main/java/io/crate/stats/ExtendedOsStats.java]: not without changing the default ctor

File: sql/src/main/java/io/crate/stats/ExtendedProcessCpuStats.java (hunk 7 above)
Comment (seut) [inline sql/src/main/java/io/crate/stats/ExtendedProcessCpuStats.java]: yeah think it make sense, was copied over ;)

**Comment by mfussenegger** [comment]
ack

**Pull request merged**